tests/integration/test_named_collections/test.py

```diff
@@ -105,6 +98,139 @@ def test_access(cluster):
     assert int(node.query("select count() from system.named_collections")) > 0
 
 
+def test_granular_access_show_query(cluster):
+    node = cluster.instances["node"]
+    assert 1 == int(node.query("SELECT count() FROM system.named_collections"))
+    assert (
+        "collection1" == node.query("SELECT name FROM system.named_collections").strip()
+    )
+
+    node.query("DROP USER IF EXISTS kek")
+    node.query("CREATE USER kek")
+    node.query("GRANT select ON *.* TO kek")
+    assert 0 == int(
+        node.query("SELECT count() FROM system.named_collections", user="kek")
+    )
+
+    node.query("GRANT show named collections ON collection1 TO kek")
+    assert 1 == int(
+        node.query("SELECT count() FROM system.named_collections", user="kek")
+    )
+    assert (
+        "collection1"
+        == node.query("SELECT name FROM system.named_collections", user="kek").strip()
+    )
+
+    node.query("CREATE NAMED COLLECTION collection2 AS key1=1, key2='value2'")
+    assert 2 == int(node.query("SELECT count() FROM system.named_collections"))
+    assert (
+        "collection1\ncollection2"
+        == node.query("select name from system.named_collections").strip()
+    )
+
+    assert 1 == int(
+        node.query("SELECT count() FROM system.named_collections", user="kek")
+    )
+    assert (
+        "collection1"
+        == node.query("select name from system.named_collections", user="kek").strip()
+    )
+
+    node.query("GRANT show named collections ON collection2 TO kek")
+    assert 2 == int(
+        node.query("SELECT count() FROM system.named_collections", user="kek")
+    )
+    assert (
+        "collection1\ncollection2"
+        == node.query("select name from system.named_collections", user="kek").strip()
+    )
+    node.restart_clickhouse()
+    assert (
+        "collection1\ncollection2"
+        == node.query("select name from system.named_collections", user="kek").strip()
+    )
+
+    node.query("DROP USER IF EXISTS koko")
+    node.query("CREATE USER koko")
+    node.query("GRANT select ON *.* TO koko")
+    assert 0 == int(
+        node.query("SELECT count() FROM system.named_collections", user="koko")
+    )
+    node.query("GRANT show named collections ON * TO koko")
+    assert (
+        "collection1\ncollection2"
+        == node.query("select name from system.named_collections", user="koko").strip()
+    )
+    node.restart_clickhouse()
+    assert (
+        "collection1\ncollection2"
+        == node.query("select name from system.named_collections", user="koko").strip()
+    )
+
+    node.query("DROP NAMED COLLECTION collection2")
+
```

Comment:
What it will show for the following case:

```
GRANT CREATE NAMED COLLECTION ON name1 TO kek;
GRANT SELECT ON name1.* TO kek;
SHOW GRANTS FOR kek;
```

There is nothing wrong with having a named collection and a table with the same name. Will it show the rights as
```
GRANT CREATE NAMED COLLECTION, SELECT ON name1.* TO kek
```
or
```
GRANT CREATE NAMED COLLECTION, SELECT ON name1 TO kek
```
?

---

Another question - what about this:
```
GRANT CREATE NAMED COLLECTION ON * TO kek;
GRANT SELECT ON * TO kek;
SHOW GRANTS FOR kek;
```

Here `GRANT SELECT ON * TO kek` just means `GRANT SELECT ON <current_database>.* TO kek`.

Reply:
I think we need a test for that.

Reply:
added those cases to the test too